Add common stmemsc routines

avisconti opened the pull request:
This PR is a temptative to solve issue #33440
The PR introduces new common i2c/spi read/write routines which may be used by all ST sensor drivers that are based on STMEMSC HAL i/f (see https://github.com/zephyrproject-rtos/hal_st/tree/master/sensor/stmemsc).
 
 To test the common routines the iis2iclx sensor has been used. This because the LSM6DSO driver, which was the object of the request, has to be first modified into a multi-instance driver using the DT helpers. After PR review I will proceed to convert LSM6DSO in multi-istance and use these new stmemsc routines. This just for my convenience.                                                                                
                                                                                        
Tested on x_nucleo_iks02a1 shield:                                                                                
    - IIS2ICLX istance on I2C (on shield)                                                                                
    - IIS2ICLX instance on SPI (on DIL24)                                                                              
                                                                     
EDIT:
This PR IS NOT solving the issue! Still, a patch to the i2c write routine has to be applied to avoid the burst write for Nordic platforms, as done for LIS2DH sensor in #19901

galak commented on include/drivers/sensor/stmemsc.h:
```diff
@@ -0,0 +1,37 @@
+ * zephyrproject-rtos/modules/hal/st/sensor/stmemsc/
+ */
+
+#ifndef ZEPHYR_DRIVERS_SENSOR_STMEMSC_STMEMSC_H_
```
This would be better placed in drivers/sensor/stmemsc

galak reviewed (changes requested):
Looks good except the placement of the header file.

avisconti committed:
drivers/sensor: iis2iclx: Make use of stmemsc common routines

Make use of the stmemsc i2c/spi read/write routine that
has been introduced to simplify the ST sensor drivers code.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
drivers/sensor: stmemsc: Add common i2c/spi read/write routines

Add common i2c/spi read/write routines for the benefit of those
ST sensor drivers who make use of the stmemsc HAL i/f.

Add generic stmemsc_cfg_i2c and stmemsc_cfg_spi structures which
contains all relevant information required by i2c/spi low level
routines, such as:

    - the pointer to the bus device
    - the I2C slave address (if instance is on I2C bus)
    - the spi_config structure (if istance is on SPI bus)

This level of abstraction allows the re-use of the i2c/spi read/write
routines among all stmemsc based sensor driver without the need to
reference the specific sensor data and or config structures.

The STMEMSC HAL source code is located here:
     zephyrproject-rtos/modules/hal/st/sensor/stmemsc/

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
> Looks good except the placement of the header file.

OK moved under drivers/sensor/stmemsc. Moreover I noticed that the first commit was carrying a part of the second so I split it, and squashed the correct part with the second (stmemsc routines)

avisconti commented:
I marked this PR as bug, since that, indirectly, it aims to solve #33440.

galak commented on drivers/sensor/stmemsc/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+zephyr_library_sources_ifdef(CONFIG_I2C        stmemsc_i2c.c)
+zephyr_library_sources_ifdef(CONFIG_SPI        stmemsc_spi.c)
+
+zephyr_include_directories(./)
```
Is it possible to not add this?

ie, can we do `#include "../stmemsc/stmemsc.h" in the actual drivers?

avisconti replied:
> Is it possible to not add this?
> 
> ie, can we do `#include "../stmemsc/stmemsc.h" in the actual drivers?

Do you think it is better? Of course I don't see any practical issue in doing what you suggest, but it seems less elegant and I would like to understand if there is a reason to not add the stmemsc directory in the include PATH. It seems to me more convenient than creating a dependence in all the drivers (at the end will be many) that will include stmemsc.h

avisconti replied:
@galak 
I have done the requested modification. Still, before submitting it I would like, if possible, to get a rational why should be better in this way. Is there any other already existing example in the source code? Thanks for any comment on this.

OPinions from the other reviewers?

galak replied:
The concern is that when we add `zephyr_include_directories` this gets added globally for all files that get built.  So I feel like the more we can limit those cases the better.

@tejlmand do we have any means to limit the scope of `include_dirs` getting added only for specific source files?

avisconti replied:
@galak I re-pushed following your suggestions

galak replied:
thanks

mbolivar-nordic replied:
> @tejlmand do we have any means to limit the scope of `include_dirs` getting added only for specific source files?

Did you try using zephyr_library_include_directories()?

avisconti replied:
@mbolivar-nordic @galak @tejlmand 
To be honest I do not like the <#include "../stmemsc/stmemsc.h"> approach. I would prefer having a different solutions. Now my question is what is the difference between the two macros (i.e. zephyr_library_include_directories() vs zephyr_include_directories()) and what is anyway the concern behind it?
I have to re-push anyway, so it would be good to find the better way to do this stuff.
Thanks for giving me your valuable opinions.

galak commented on drivers/sensor/stmemsc/stmemsc_i2c.c:
```diff
@@ -0,0 +1,24 @@
+ * zephyrproject-rtos/modules/hal/st/sensor/stmemsc/
+ */
+
+#include "./stmemsc.h"
```
nit: remove the `./`

avisconti replied:
ack

galak commented on drivers/sensor/stmemsc/stmemsc_spi.c:
```diff
@@ -0,0 +1,74 @@
+ * zephyrproject-rtos/modules/hal/st/sensor/stmemsc/
+ */
+
+#include "./stmemsc.h"
```
remove `./`

avisconti replied:
ack

galak reviewed (commented):
lgtm, minor tweak on the include use of `./`

avisconti commented:
- Coding style -
Rewritten the SPI routines in a more compact way, but w/o changing the actual implementation.
Re-tested on x_nucleo_iks01a2 (2 iis2iclx sensors - I2C and SPI)

MaureenHelm reviewed (approved):
This is a nice simplification

mbolivar-nordic reviewed (commented):
Not blocking, but using 96 ~bytes~ bits for every I2C device's `stmemsc_bus_cfg` when it could be 16 as it is now seems wasteful.

edit: fixed units

avisconti commented on drivers/sensor/stmemsc/stmemsc_spi.c:
```diff
@@ -0,0 +1,68 @@
+
+#include "stmemsc.h"
+
+#define IIS2ICLX_SPI_READ		(1 << 7)
```
oooops! Thanks for catching it! 
It's a copy/paiste issue: I'm particullarly used to this kind of errors! I'll correct asap.

avisconti commented:
Re-tested and re-pushed again. THanks very much for all the reviews I received and nice suggestions.
Summary:
1. Used SPI_READ instead of IIS2ICLX_SPI_READ tha was a typo (thanks @erwango).
2. Remove the uinion stmesmc_bus_cfg. Instead, I added two different structures, stmesc_cfg_i2c and stmemsc_cfg_spi, and I include them in the device driver using the stmemsc common routines based on DT_ANY_INST_ON_BUS_STATUS_OKAY(). This means that: 
2.A if a particular application is using instances on both bus will be slightly worse than before in the size of its DEVX_CFG structure (before there was a union)
2.B if instances are on SPI bus only will be same as before in size
2.C if instances are on I2C bus only the size will be drammatically reduced!
(Thanks @mbolivar-nordic )
3. The  <#include "../stmemsc/stmemsc.h"> approach is still there. I don't like it, but I'm waiting for a better suggestion (maybe using zephyr_library_include_directories()?). Please comment on that.

erwango commented on drivers/sensor/iis2iclx/iis2iclx.h:
```diff
@@ -15,6 +15,7 @@
 #include <zephyr/types.h>
 #include <drivers/gpio.h>
 #include <sys/util.h>
+#include "../stmemsc/stmemsc.h"
```
What about moving `stmemsc.h` to `include/drivers/sensor`and include it from there ?

erwango replied:
ok, it seems this was the initial proposal. Though I'm not sure to understand why this is not a valid solution

avisconti replied:
Correct, it was the initial proposal. To me was the best solution. In that case no need to add an include path.

tejlmand replied:
you should be able to do:
```suggestion
#include <stmemsc.h>
```

and then in: https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/sensor/iis2iclx/CMakeLists.txt
```
zephyr_library_include_directories(../stmemsc)
```

avisconti replied:
OK, I will do like that. Thanks for suggestion.

mbolivar-nordic commented on drivers/sensor/iis2iclx/iis2iclx_i2c.c:
```diff
@@ -10,42 +10,24 @@
 
 	data->ctx = &data->ctx_i2c;
-	data->ctx->handle = (void *)dev;
+	data->ctx->handle = (void *)&cfg->stmemsc_i2c;
```
This is not related to this PR, but I'm a bit confused about the `ctx`, `ctx_i2c`, and `ctx_spi` members in `iis2iclx_data`.

I see that the HAL functions take an `stmdev_ctx_t*` instead of a `const stmdev_ctx_t*`, so I think I understand why this is allocated in RAM (though it's a pity because it seems like it could maybe go into ROM).

However, I don't understand why `iis2iclx_data` is allocating all three `ctx`, `ctx_i2c`, and `ctx_spi`. Why not just have:

```c
struct iis2iclx_data {
	stmdev_ctx_t ctx;
};
```

Then here (and similarly in the SPI file):

```c
data->ctx.read_reg = (stmdev_read_ptr)stmemsc_i2c_read;
data->ctx.write_reg = (stmdev_read_ptr)stmemsc_i2c_write;
data->ctx.handle = &cfg->bus_union_i_suggested_earlier.stmemsc_i2c;
```

And use `&data->ctx` wherever `data->ctx` appears today?

I feel this driver may be a basis for a regmap framework someday so I'm just asking more general questions about memory optimization.

avisconti replied:
> This is not related to this PR, but I'm a bit confused about the `ctx`, `ctx_i2c`, and `ctx_spi` members in `iis2iclx_data`.
> 
It is not related to this PR, but I feel that you are sorting up some confusion I had about multi-instance.                                                                                                     

> I see that the HAL functions take an `stmdev_ctx_t*` instead of a `const stmdev_ctx_t*`, so I think I understand why this is allocated in RAM (though it's a pity because it seems like it could maybe go into ROM).                                                                                                     
                                                                                                     
In this case I prefer to keep this part in data, and let the fill be done in the init function (either I2C or SPI) set in configuration. This just because I'm not sure whether the interface would remain always the same and I prefer leave that part at runtime.                                                                                                     

> 
> However, I don't understand why `iis2iclx_data` is allocating all three `ctx`, `ctx_i2c`, and `ctx_spi`. Why not just have:
> 
> ```c
> struct iis2iclx_data {
> 	stmdev_ctx_t ctx;
> };
> ```
> 
> Then here (and similarly in the SPI file):
> 
> ```c
> data->ctx.read_reg = (stmdev_read_ptr)stmemsc_i2c_read;
> data->ctx.write_reg = (stmdev_read_ptr)stmemsc_i2c_write;
> data->ctx.handle = &cfg->bus_union_i_suggested_earlier.stmemsc_i2c;
> ```
> 
> And use `&data->ctx` wherever `data->ctx` appears today?
>                                                                                                      
Yes, of course. The driver structure is done in such a way we have a unique couple of (config, data) structures for each instance and stmdev_ctx_t is general, so no need for allocating separate structures for I2C and SPI. I think that initially (long back ago) the stmdev_ctx_t structure was meant to be different between the busses, and by lazyness I never went through this implementation again since that. Embarassing...

> I feel this driver may be a basis for a regmap framework someday so I'm just asking more general questions about memory optimization.

Sure. It will be the basis for a general re-structuring of stmemsc based sensor at least. Thanks for spending time on that.

avisconti replied:
I feel I should incorporate all the suggestions in this PR

mbolivar-nordic replied:
>  It will be the basis for a general re-structuring of stmemsc based sensor at least.

The journey of a thousand miles begins with a single step. Thanks for a very nice cleanup.

mbolivar-nordic reviewed (commented):
Some non-blocking questions and observations.

avisconti commented on drivers/sensor/stmemsc/stmemsc_spi.c:
```diff
@@ -0,0 +1,68 @@
+	};
+	const struct spi_buf_set rx = { .buffers = rx_buf, .count = 2 };
+
+	if (spi_transceive(stmemsc->bus, spi_cfg, &tx, &rx)) {
```
Agree

avisconti commented on drivers/sensor/iis2iclx/iis2iclx.h:
```diff
@@ -35,20 +36,14 @@
-	struct spi_config spi_cfg;
-#endif /* DT_ANY_INST_ON_BUS_STATUS_OKAY(spi) */
-};
-
-struct iis2iclx_config {
-	const struct device *bus;
+	const struct stmemsc_cfg_spi stmemsc_spi;
```
Something like: 

``` 
union {
#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
    const struct stmemsc_cfg_i2c stmemsc_i2c;
#endif
#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
    const struct stmemsc_cfg_spi stmemsc_spi;
#endif
} cfg;

avisconti committed:
drivers/sensor: iis2iclx: avoid using both ctx_i2c and ctx_spi

The STMEMSC HAL i/f requires every driver using it to define
for each instance a stmdev_ctx_t structure, which is used to
export the hardware specific APIs to handle i2c and spi busses
operations. Since this structure is bus agnostic, there is no
need to declare it twice for both i2c and spi. Instead, declare
only one structure, which will be populated either with i2c APIs
or with spi APIs according to where that particular instance is
declared inside the DT.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
drivers/sensor: iis2iclx: Fix sensorhub building

The iis2iclx driver was not building correctly when
CONFIG_IIS2ICLX_SENSORHUB was enabled.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
THis should probably be the last re-push. I also fixed a compilation error when enabling sensorhub and tested the sensorhub part using x_nucleo_iks02a1 shield configured in sensorhub mode.
